src/main/scala/sigmastate/serialization/ErgoScriptSerializer.scala

```diff
@@ -0,0 +1,28 @@
+package sigmastate.serialization
+
+import sigmastate.SType
+import sigmastate.Values.{Constant, Value}
+import sigmastate.eval.RuntimeIRContext
+
+object ErgoScriptSerializer {
+  def serialize(script: Value[SType]): Array[Byte] = ???
+  def deserialize(bytes: Array[Byte]): Value[SType] = ???
+
+  def extractConstants(script: Value[SType]): (Seq[Constant[_]], Value[SType]) = {
+    val IR: RuntimeIRContext = new RuntimeIRContext {}
+    import IR._
+    implicit val eAny = stypeToElem(script.tpe).asInstanceOf[Elem[Context]]
+
+    val env = Map[String, Any]()
+    val IR.Pair(calcF, _) = doCosting(env, script)
+
+    val f = fun { in: Rep[Context] =>
```

Comment:
@ergomorphic What's the best approach to traversing the script tree? I'm thinking of building the graph out of it and implement the traversing and constant swapping in `fun` wrapped around the `calcF` and passing it to the `buildTree`. I feel like there should be an easier way.

Reply:
@greenhat You don't need to do new traversal of the script tree. I think the constant replacement should be done in TreeBuilding.buildTree. see `case Def(Const(x)) =>`. Instead of mkConstant you need to do mkPlaceholder(nPlaceHolder). Keep track of nPlaceHolder counters and constants. 
Currently all constants in Scalan graphs are unified, this should be changed for our specific use case. It is not hard to do. So you can assume that different constants subexpressions in source script are different IR nodes in Scalan graph.